idaes/apps/grid_integration/multiperiod/price_taker_model.py

```diff
@@ -0,0 +1,968 @@
+              using representative days
+        - Add documentation to the IDAES readthedocs
+        - Add a warning when costs/revenues are given,
+          but are never used during cashflow construction
+"""
+import pandas as pd
+import numpy as np
+from functools import reduce
+from operator import attrgetter
+import os.path
+
+from importlib import resources
+from pathlib import Path
+from pyomo.environ import (
+    ConcreteModel,
+    Block,
+    Var,
+    RangeSet,
+    Objective,
+    Constraint,
+    NonNegativeReals,
+    Expression,
+    maximize,
+    Param,
+)
+
+from idaes.apps.grid_integration import MultiPeriodModel
+from idaes.apps.grid_integration.multiperiod.design_and_operation_models import (
+    DesignModelData,
+    OperationModelData,
+)
+
+from sklearn.cluster import KMeans
+from kneed import KneeLocator
+
+import matplotlib.pyplot as plt
+
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class PriceTakerModel(ConcreteModel):
+    def __init__(self, seed=20, horizon_length=24):
+        super().__init__()
+        self.seed = seed
+        self.horizon_length = horizon_length
+
+    @property
+    def seed(self):
+        return self._seed
+
+    @seed.setter
+    def seed(self, value):
+        if not isinstance(value, int):
+            raise ValueError(f"seed must be an integer, but {value} is not an integer")
+        self._seed = value
+
+    @property
+    def horizon_length(self):
+        return self._horizon_length
+
+    @horizon_length.setter
+    def horizon_length(self, value):
+        if value <= 0:
+            raise ValueError(f"horizon_length must be > 0, but {value} is provided.")
+        if not isinstance(value, int):
+            raise ValueError(
+                f"horizon_length must be an integer, but {value} is not an integer"
+            )
+        self._horizon_length = value
+
+    def generate_daily_data(self, raw_data):
+        """
+        Function used to generate the daily data in a usable format
+        from the raw data provided.
+
+        Args:
+            raw_data:   Columnar data for a given LMP signal
+
+        Returns:
+            daily_data: Correctly formatted daily LMP data for later use
+        """
+        day_list = list(range(1, (len(raw_data) // self._horizon_length) + 1))
+
+        daily_data = pd.DataFrame(columns=day_list)
+
+        # Extracting data to populate empty dataframe
+        i = 0
+        j = self._horizon_length
+        day = 1
+
+        if j > len(raw_data):
+            raise ValueError(
+                f"tried to generate daily data, but horizon length of {self._horizon_length} exceeds raw_data length of {len(raw_data)}"
+            )
+
+        while j <= len(raw_data):
+            daily_data[day] = raw_data[i:j].reset_index(drop=True)
+            i = j
+            j = j + self._horizon_length
+            day = day + 1
+
+        return daily_data
+
+    def get_optimal_n_clusters(
```

Comment:
@lbianchi-lbl I was wondering if you noticed this thread and had any thoughts about bounds on scikitlearn version supported in IDAES. For example, @djlaky points out that the optimal number of clusters in his test case can be either 10 or 11, depending on whether older or newer versions of scikit-learn are used. I am unsure which versions @djlaky is referring to when he describes them as "older" or "newer" versions, but it might be good to consider a solution to this issue to help merge this Pricetaker model.